### DISCUSSION TITLE
Pdf streaming

### DISCUSSION THREAD

**Discussion by dhananjay19100** [stephanrauh/ngx-extended-pdf-viewer#2014] [Q&A]
I have uploaded a few PDFs to an S3 bucket, and the PDF viewer is downloading some PDFs in chunks while others are downloaded in full before being displayed. The metadata of the PDFs and the server are also the same. Can you explain why this is happening?

pdf which is download full: https://pdf-testing.s3.ap-south-1.amazonaws.com/2gb-sample-file.pdf

**Comment by stephanrauh** [comment]
According to my documentation (https://pdfviewer.net/extended-pdf-viewer/range-requests), you need three response headers:

* "Content-Length"
* "Accept-Ranges": "bytes",
* "Content-Encoding"

Your S3 file only returns two of these three attributes. I'm not sure why `Content-Encoding` is necessary, but I trust I've written this because of a reason, even though I've forgotten the reason in the meantime.

**Reply by dhananjay19100** [reply]
https://fs.dxassist.ai/01HDDEXN03FZ9PZ2S7MSA2BSZX?Expires=1705208866&Signature=MsIHzhQ1l0nYGId6xFGVF7FMc7tfKTCo-3v4EZVCD5kx73-RCBy0IOuGJheReQZMvV97b6~BPJPZIIyPajnwZ-ka8~j2tIOZe76pkGs7~0gfsnYZkbljucyc4xVS-8wA3y9o3vqBfilUsdtJYRGqyV5ZQvq~~dWgmGg5YqlGVD6yaJv5XgDHIyC1qdBLdBTZ-mUXq40JCPdKFlD~6iVJYjOctNdSIyD~ULVGGCUI90Nu98nFJUJeats33bgK-azQr~EVRj9wnOXKjtyM7TaK~KbJbfM1~BnPRZgUIhg6BU0Ak4ZJCieagJRjwb9Qz91OBpma1eVcod3L3U-fBK7HPg__&Key-Pair-Id=K2BC7R8SEVH8JU  
this is another pdf url which is streaming and this pdf dons't have Content-Encoding.

**Reply by stephanrauh** [reply]
This time, `Accept-Ranges` seems to return null. You can debug this yourself. Set `minifiedJSLibraries="false"`, open the file `pdf-xx.js` in your browser's debugger, navigate to the method `validateRangeRequestCapabilities()`, and set a breakpoint.

```js
function validateRangeRequestCapabilities({
  getResponseHeader,
  isHttp,
  rangeChunkSize,
  disableRange,
}) {
  if (typeof PDFJSDev === "undefined" || PDFJSDev.test("TESTING")) {
    assert(
      Number.isInteger(rangeChunkSize) && rangeChunkSize > 0,
      "rangeChunkSize must be an integer larger than zero."
    );
  }
  const returnValues = {
    allowRangeRequests: false,
    suggestedLength: undefined,
  };

  const length = parseInt(getResponseHeader("Content-Length"), 10);
  if (!Number.isInteger(length)) {
    return returnValues;
  }

  returnValues.suggestedLength = length;

  if (length <= 2 * rangeChunkSize) {
    // The file size is smaller than the size of two chunks, so it does not
    // make any sense to abort the request and retry with a range request.
    return returnValues;
  }

  if (disableRange || !isHttp) {
    return returnValues;
  }
  if (getResponseHeader("Accept-Ranges") !== "bytes") {
    return returnValues;
  }

  const contentEncoding = getResponseHeader("Content-Encoding") || "identity";
  if (contentEncoding !== "identity") {
    return returnValues;
  }

  returnValues.allowRangeRequests = true;
  return returnValues;
}
```

**Reply by dhananjay19100** [reply]
how i can run your code in my system

**Reply by stephanrauh** [reply]
Please elaborate. Which code do you want to run, and why?

**Reply by dhananjay19100** [reply]
to check validateRangeRequestCapabilities() methode for chunking

**Reply by stephanrauh** [reply]
That's not my code. It's the source code of pdf.js.

**Comment by dhananjay19100** [comment]
In you new version 19.0.3, '[(page)]="pageNumber" method is not working

**Reply by dhananjay19100** [reply]
it always open 1st page

**Reply by stephanrauh** [reply]
What? But you're right, my showcase demo almost always shows page 1 instead of page 5. Weird. I've opened a dedicated ticket.
